### PR TITLE
Rework windows font collector

### PR DESCRIPTION
**Pull request by moi15moi**
The main goal of this PR is to replace the "hack" in the Windows FontCollector which involves using the Windows registry to obtain the font filename. Now, we use DirectWrite which can interop with GDI, so we are 100% sure it use the same font has VSFilter.

I have also removed Uniscribe, which was used to determine if a font has a specific character.
The sole objective of this PR is to enhance the code quality in the Windows FontCollector.

Now, the user can request a font by a fullname or postscript name like said in https://github.com/libass/libass/pull/203. Previously, on windows, the user could only request a family name.

Should I move the method ``normalizeFilePathCase`` to ``Aegisub/libaegisub/windows/fs.cpp``?

Edit:
The commit de1d3e38d19b1fa3a762450bb3c59bc4e13f7e23 ensures that fonts installed via [AddFontResource](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew) are detected by FontCollector. Previously, they were not recognized.

However, please note that this detection only works on Windows 10 or later. It's uncertain whether this behavior is a bug in the Windows API, but the fonts installed via [AddFontResource](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew) aren't added to the IDWriteFontCollection. Consequently, we cannot convert IDWriteFontFace to IDWriteFont which impeding the ability to verify the presence of certain characters.

### CODE REVIEW

**Comment by moi15moi** [comment]
The CI failure for MacOS seems unrelated to this PR.
```
/usr/bin/hdiutil detach /dev/disk2 -force
hdiutil: couldn't eject "disk2" - Resource busy
FAILED: meson-internal__osx-build-dmg 
env MESON_SOURCE_ROOT=/Users/runner/work/Aegisub/Aegisub MESON_BUILD_ROOT=/Users/runner/work/Aegisub/Aegisub/build MESON_SUBDIR=packages 'MESONINTROSPECT=/Users/runner/hostedtoolcache/Python/3.12.1/x64/bin/meson introspect' /Users/runner/work/Aegisub/Aegisub/tools/osx-dmg.sh /Users/runner/work/Aegisub/Aegisub /Users/runner/work/Aegisub/Aegisub/build 3.2.2
```

File: src/font_file_lister_gdi.cpp [hunk 1]
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
...
+	agi::scoped_holder<WCHAR**> path_sh(&path, [](WCHAR** p) { delete[] *p; });
+
+	if (FAILED(local_loader_sh->GetFilePathFromKey(font_file_reference_key, font_file_reference_key_size, path, path_length + 1)))
+	{
+		return ret;
+	}
+
+	// DirectWrite always return the filepath in upper case. Ex: "C:\WINDOWS\FONTS\ARIAL.TTF"
+	DWORD normalized_path_length;
+	if (!normalizeFilePathCase(&path, &normalized_path_length)) {
```
Comment (arch1t3cht) [inline src/font_file_lister_gdi.cpp:279]: It'd be more consistent to use `path_sh` rather than `&path` here (and similarly in the `ag::fs::path` call below).

**Review by arch1t3cht — CHANGES_REQUESTED** [review]
I finally had time to look through this - thanks a lot for the PR!

Looking through it, everything seems to be correct as long as I trust you on the DirectWrite/GDI wrangling (and barring the `IDWriteLocalFontLoader` question on Discord). I do have a few comments regarding general code style and conventions, though.

File: meson.build [hunk 2]
```diff
@@ -327,6 +327,10 @@ if host_machine.system() == 'windows'
     endif
 endif
 
+if host_machine.system() == 'windows' and cc.has_header('dwrite_3.h')
+    add_project_arguments('-DHAVE_DWRITE_3', language: 'cpp')
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply meson.build]: Addressed in commit 7d099da0994ec37d3584e6c9ce4242f6555f7e8e

File: src/dialog_fonts_collector.cpp [hunk 3]
```diff
@@ -94,7 +94,16 @@ void FontsCollectorThread(AssFile *subs, agi::fs::path const& destination, FcMod
...
 
-		auto paths = FontCollector(AppendText).GetFontPaths(subs);
+		std::vector<agi::fs::path> paths;
+		try {
+			paths = FontCollector(AppendText).GetFontPaths(subs);
+		}
+		catch (agi::EnvironmentError const& err) {
+			AppendText(fmt_tl("* An error occurred when collecting the font: %s.\n", err.GetMessage()), 2);
+			collector->AddPendingEvent(wxThreadEvent(EVT_COLLECTION_DONE));
+			return;
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/dialog_fonts_collector.cpp]: Addressed in commit 969758825d452fe347202b91b8b633c1e4925235

File: src/font_file_lister_gdi.cpp [hunk 4]
```diff
@@ -16,168 +16,130 @@
...
+
+/// @brief Normalize the case of a file path.
+///        Ex: For "C:\WINDOWS\FONTS\ARIAL.TTF", it would return "C:\Windows\Fonts\arial.ttf"
+/// @param[out] path   The path to be normalized. It can be a directory or a file.
+/// @param[out] length Length of the normalized file path string, not including the terminated NULL character.
+/// @return Returns a boolean value indicating the success of the normalization process.
+///          - Returns true if the normalization is successful.
+///          - Returns false if an error occurs during file handling or path normalization.
+bool normalizeFilePathCase(WCHAR** path, DWORD* length) {
+	if (path == nullptr || length == nullptr) {
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I don't use WCHAR anymore, so it make this remark obsolete.

File: src/font_file_lister_gdi.cpp [hunk 5]
```diff
@@ -16,168 +16,130 @@
...
+		return false;
+	}
 
-	// System fonts
-	read_fonts_from_key(HKEY_LOCAL_MACHINE, font_dir, files);
+	// GetFinalPathNameByHandle return path into ``device path`` form. Ex: "\\?\C:\Windows\Fonts\ariali.ttf"
+	// We need to convert it to ``fully qualified DOS Path``. Ex: "C:\Windows\Fonts\ariali.ttf"
+	// There isn't any public API that remove the prefix (there is RtlNtPathNameToDosPathName, but it is really hacky to use it)
+	// See: https://stackoverflow.com/questions/31439011/getfinalpathnamebyhandle-result-without-prepended
+	// Even cython remove the prefix manually: https://github.com/python/cpython/blob/963904335e579bfe39101adf3fd6a0cf705975ff/Lib/ntpath.py#L733-L793
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 5212d1c7ef686de30f863de0f72cb57208f851c0

File: src/font_file_lister_gdi.cpp [hunk 6]
```diff
@@ -16,168 +16,130 @@
...
+#include <windowsx.h>
+
+#ifdef HAVE_DWRITE_3
+#include <dwrite_3.h>
+#endif
+
+/// @brief Normalize the case of a file path.
+///        Ex: For "C:\WINDOWS\FONTS\ARIAL.TTF", it would return "C:\Windows\Fonts\arial.ttf"
+/// @param[out] path   The path to be normalized. It can be a directory or a file.
+/// @param[out] length Length of the normalized file path string, not including the terminated NULL character.
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: > Or, even better, the input and output could be replaced with some higher-level C++ class like `std::wstring`.

Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I use ``std::wstring``.

File: src/font_file_lister_gdi.cpp [hunk 7]
```diff
@@ -16,168 +16,130 @@
...
+	int weight = bold == 0 ? 400 :
+				 bold == 1 ? 700 :
+				 bold;
+
+	// From VSFilter
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/RTS.cpp#l45
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l2992
 	LOGFONTW lf{};
-	lf.lfCharSet = DEFAULT_CHARSET;
+	lf.lfCharSet = DEFAULT_CHARSET; // we should use the one specified in the ass file
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp:104]: I wanted to talk about that. In some very specific situation, if we don't use the charset requested by the user, the selected font can be different then the font displayed by VSFilter.
But, this can hardly be emulated on macOS and Linux. So, do we really care to use the charset specified by the user? Even libass doesn't properly emulate how the charset works: https://github.com/libass/libass/issues/662

File: src/font_file_lister_gdi.cpp [hunk 8]
```diff
@@ -16,168 +16,130 @@
...
+	{
+		throw agi::EnvironmentError("Failed to initialize the Gdi Interop");
+	}
+	gdi_interop_sh = gdi_interop;
 }
 
 CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, int bold, bool italic, std::vector<int> const& characters) {
 	CollectionResult ret;
 
+	int weight = bold == 0 ? 400 :
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp:96]: Simply for clarity.

File: src/font_file_lister_gdi.cpp [hunk 9]
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
...
-			continue;
-		if (memcmp(&file_buffer[0], &buffer[0], buffer.size()) == 0) {
-			ret.paths.push_back(it->second);
-			break;
-		}
+	HFONT hfont = CreateFontIndirect(&lf);
+	if (hfont == nullptr) {
+		return ret;
 	}
+	agi::scoped_holder<HFONT> hfont_sh(hfont, [](HFONT p) { DeleteObject(p); });
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 9af7c22052d59c884022c4875304ee6d9cfd722e

File: src/font_file_lister_gdi.cpp [hunk 10]
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
...
-		}
+	IDWriteFontFace* font_face;
+	if (FAILED(gdi_interop_sh->CreateFontFaceFromHdc(dc_sh, &font_face)))
+	{
+		return ret;
 	}
+	agi::scoped_holder<IDWriteFontFace*> font_face_sh(font_face, [](IDWriteFontFace* p) { p->Release(); });
+
+#ifdef HAVE_DWRITE_3
+	// Fonts added via the AddFontResource API are not included in the IDWriteFontCollection.
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp:164]: No, the old font collector wouldn't find them since the fonts added by AddFontResource aren't registered in the Windows Registry.

File: src/font_file_lister_gdi.cpp [hunk 11]
```diff
@@ -16,168 +16,130 @@
...
-		}
+GdiFontFileLister::GdiFontFileLister(FontCollectorStatusCallback &)
+: dwrite_factory_sh(nullptr, [](IDWriteFactory* p) { p->Release(); })
+, font_collection_sh(nullptr, [](IDWriteFontCollection* p) { p->Release(); })
+, dc_sh(nullptr, [](HDC dc) { DeleteDC(dc); })
+, gdi_interop_sh(nullptr, [](IDWriteGdiInterop* p) { p->Release(); })
+{
+	IDWriteFactory* dwrite_factory;
+	if (FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), reinterpret_cast<IUnknown**>(&dwrite_factory))))
+	{
```
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 863d42723d2790aa96330b899e1e20da63c89bf8

File: src/font_file_lister_gdi.cpp (hunk 1 above)
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I don't use WCHAR anymore, so it make this remark obsolete.

**Comment by moi15moi** [comment]
> Looking through it, everything seems to be correct as long as I trust you on the DirectWrite/GDI wrangling (and barring the `IDWriteLocalFontLoader` question on Discord).

I did a bunch of test with the user.
You can see my conclusion [here](https://github.com/moi15moi/FindSystemFontsFilename/commit/890301cb79f3c27df5d0a0d04a77ef9349db3900).

File: src/font_file_lister_gdi.cpp [hunk 12]
```diff
@@ -210,90 +149,85 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
...
+	LPCVOID font_file_reference_key;
+	UINT32 font_file_reference_key_size;
+	if (FAILED(font_file_sh->GetReferenceKey(&font_file_reference_key, &font_file_reference_key_size)))
+		return ret;
+
+	UINT32 path_length;
+	if (FAILED(local_loader_sh->GetFilePathLengthFromKey(font_file_reference_key, font_file_reference_key_size, &path_length)))
+		return ret;
+
+	WCHAR* path = new WCHAR[path_length + 1];
```
Comment (moi15moi) [inline src/font_file_lister_gdi.cpp:221]: I forgot to release this variable.
I will need to use agi::scoped_holder.

Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 62813a6650fe5d645aaf29048d07bd9586662839

**Comment by moi15moi** [comment]
In some specific cases, the FontCollector can make a mistake in obtaining the font weight and italic style.
For example, install "AliviaRegular_Weight31961.ttf" and "Alivia.otf," which you can find [here](https://github.com/libass/libass/issues/437#issuecomment-1925944380) in the .zip file.

FontCollector will return the font path to "AliviaRegular_Weight31961.ttf" (which is correct), but it will select the weight and italic style from "Alivia.otf." Because of this problem, the ``fake_bold`` and ``fake_italic`` doesn't always represent the true information.

I tried to see if there is a way to avoid this problem, but I didn't find one.
Here is what I tried:

1. Always using the first logfont returned by EnumFontFamiliesEx. I thought the first logfont returned by the Enum is the one that would be selected in the HDC, but it isn't the case. A simple test with Arial shows that it won't necessarily use the first logfont.
2. GetTextMetricsW and GetOutlineTextMetricsW. They return the "simulated" font. For example, if you ask for a font with weight 700, but there is only a 400 font installed, both methods will return metrics with weight 700.
3. IDWriteFont::GetWeight and IDWriteFont::GetStyle. In some specific cases, DirectWrite is incompatible with GDI, and it won't return the same italic value. The same problem occurs with the weight. See this issue: https://github.com/moi15moi/FontCollector/issues/8#issuecomment-1755885113

Conclusion:
There isn't any function that GDI offers to know the italic and weight of the selected font in the HDC.
The only "perfect" solution would be to parse the font like GDI. We could use IDWriteFontFace::TryGetFontTable and get the OS/2 table. If a font doesn't contain an OS/2 table, we would need to use the head table. I don't know if we really want to do that.

File: src/font_file_lister_gdi.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (arch1t3cht) [reply src/font_file_lister_gdi.cpp:104]: Yes, it's probably better to treat encodings different from 1 as "unsupported". What I really meant here, though, is that currently this comment is a bit confusing. You can remove it or reword it to something like "note that this currently ignores the font encoding specified in the ass file".

File: src/font_file_lister_gdi.cpp [hunk 13]
```diff
@@ -16,284 +16,195 @@
...
+		return ret;
 
-	// User fonts
-	read_fonts_from_key(HKEY_CURRENT_USER, font_dir, files);
+	SelectFont(dc_sh, hfont_sh);
 
-	return files;
-}
+	std::wstring selected_name(LF_FACESIZE - 1, L'\0');
+	if (!GetTextFaceW(dc_sh, LF_FACESIZE, &selected_name[0]))
```
Comment (arch1t3cht) [inline src/font_file_lister_gdi.cpp:120]: Instead of `&selected_name[0]` it's better to use `selected_name.data()` here. Apart from just being more idiomatic this is technically required for correctness since `GetTextFaceW` will implicitly override the null terminator of `selected_name`, which the standard explicitly allows for `selected_name.data() + selected_name.size()` and `selected_name[selected_name.size()]` but technically not for `&selected_name[0] + selected_name.size()` (at least if I read the standard correctly. See also https://stackoverflow.com/a/41310431).

File: src/font_file_lister_gdi.cpp [hunk 14]
```diff
@@ -16,284 +16,195 @@
...
-		}
-	}
+	UINT32 path_length;
+	if (FAILED(local_loader_sh->GetFilePathLengthFromKey(font_file_reference_key, font_file_reference_key_size, &path_length)))
+		return ret;
 
-	// No fonts actually matched
-	if (ret.paths.empty())
+	std::wstring path(path_length, L'\0');
+	if (FAILED(local_loader_sh->GetFilePathFromKey(font_file_reference_key, font_file_reference_key_size, &path[0], path_length + 1)))
```
Comment (arch1t3cht) [inline src/font_file_lister_gdi.cpp:199]: Again it's better to use `path.data()` here.

File: src/font_file_lister_gdi.cpp [hunk 15]
```diff
@@ -16,284 +16,195 @@
...
-			auto stream = agi::io::Open(path, true);
-			stream->read(&buffer[0], 1024);
-			auto hash = murmur3(&buffer[0], stream->tellg());
-			hash_to_path.emplace(hash, path);
-		}
-		catch (agi::Exception const& e) {
-			cb(to_wx(e.GetMessage() + "\n"), 3);
+	// If the selected_name is different then the lf.lfFaceName,
+	// it means that the requested font doesn't exist.
+	if (_wcsnicmp(&selected_name[0], lf.lfFaceName, LF_FACESIZE))
```
Comment (arch1t3cht) [inline src/font_file_lister_gdi.cpp:125]: And then this could also be `selected_name.data()` for consistency (but this time it's just cosmetic).

**Review by arch1t3cht — COMMENTED** [review]
Thanks for the updates. Just a couple more nitpicks left. Once those are addressed, you can squash everything onto one commit and I'll merge it (or if you want to keep the commit history on your branch I'll squash it myself when merging).

File: src/font_file_lister_gdi.cpp (hunk 13 above)
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp:120]: I tried it, but I get this error:
```
../src/font_file_lister_gdi.cpp(120): error C2664: 'int GetTextFaceW(HDC,int,LPWSTR)' : impossible de convertir l'argument 3 de 'const _Elem *' en 'LPWSTR'
        with
        [
            _Elem=wchar_t
        ]
```

I'm pretty sure it is because we use [``cpp_std=c++14``](https://github.com/arch1t3cht/Aegisub/blob/cd2e894634bd6fe4038861cfc10491e4ea320123/meson.build#L4). It is only since c++17 or higher that we can use ``selected_name.data()`` in this situation. See https://stackoverflow.com/a/12742517/15835974
I guess that ``&selected_name[0]`` may depend on a undefined behaviour in this situation?

Comment (arch1t3cht) [reply src/font_file_lister_gdi.cpp:120]: Ah, yes. The `data` overload returning a non-const pointer was only added in C++17. I missed that.

However, this is a different question from what standard it'd be legal to use `data` this way in. For that, I was originally looking at cppreference, which listed C++11 for [operator[]](https://en.cppreference.com/w/cpp/string/basic_string/operator_at) and no standard requirement at all (so effectively C++17) for [data](https://en.cppreference.com/w/cpp/string/basic_string/data). But looking at the actual (final drafts of the) standards, it appears that cppreference is wrong and the clause for `operator[]` was indeed added with C++17 and the clause for `data` was added for C++20.

So as far as I can see there's no way to make this technically correct in C++11 without the "pad the string by a byte and resize it afterwards hack". However, seeing as TypesettingTools/Aegisub already switched to C++20 and since this is very likely to still work fine in practice even on earlier standards, I'd be fine with keeping this as it is and adding a comment like `FIXME: This will override the string's terminator, which is not technically correct. After switching to C++20 this should use .data().`

File: src/font_file_lister_gdi.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp]: Addressed in commit 17a3f487f2479070be45fb454c44dc3dd8d724cc

File: src/font_file_lister_gdi.cpp (hunk 13 above)
Comment (moi15moi) [reply src/font_file_lister_gdi.cpp:122]: Addressed in commit 44b81fdfae59a9096c6dcab94cf38d6d38e1052d

**Comment by moi15moi** [comment]
> Thanks for the updates. Just a couple more nitpicks left. Once those are addressed, you can squash everything onto one commit and I'll merge it (or if you want to keep the commit history on your branch I'll squash it myself when merging).

I will let you squash them. I would like to conserve the commit. Some of them contain relevant information that may be useful if a problem is discovered in the future.

**Comment by arch1t3cht** [comment]
Thanks! Now squashed and merged in [7543060](https://github.com/arch1t3cht/Aegisub/commit/7543060f1d8f008b3220850da7ad51dcb0cc11b0).
